### PR TITLE
[ADD] Reservation Workflow

### PR DESCRIPTION
**Pull request by eilst**

### CODE REVIEW

File: project_event/models/project_project.py [hunk 1]
```diff
@@ -30,6 +30,30 @@ class Project(models.Model):
...
     )
+    state = fields.Selection([
+        ('draft', 'Draft'),
+        ('option', 'Option'),
+        ('requested', 'Requested'),
+        ('read', 'Read'),
+        ('accepted', 'Accepted'),
+        ('done', 'Done'),
+        ('canceled', 'Canceled')],
+        string='Task State',
```
Comment (jananjoy) [inline project_event/models/project_project.py]: @eilst ;"], " bring in the new line .
project_project.py mostly handle for event information and why you define the string as string='Task State',
Please order the selection fields type based on chronological order

File: project_event/models/project_task.py [hunk 2]
```diff
@@ -189,12 +194,36 @@ def get_equipment_ids_inside(self):
         room_id = self.env['resource.calendar.room'].browse(self.room_id).id
         return room_id.instruments_ids.ids
 
+    @api.multi
+    def cancel_resources_reservation(self):
+        reservation_event = self.env['calendar.event'].browse(
```
Comment (jananjoy) [inline project_event/models/project_task.py]: @eilst Please format the code for better readability e.g
 reservation_event = self.env['calendar.event'].\
browse(
e.g  def create(self, vals): in project_task model

File: project_event/models/project_task.py [hunk 3]
```diff
@@ -189,12 +194,36 @@ def get_equipment_ids_inside(self):
...
+            {
+                'state': 'cancelled'
+            }
+        )
+
+    @api.multi
+    def draft_resources_reservation(self):
+        if not self.reservation_event_id:
+            self.request_reservation()
+        reservation_event = self.env['calendar.event'].browse(
```
Comment (jananjoy) [inline project_event/models/project_task.py:216]: @eilst Please format the code for better readability e.g
reservation_event = self.env['calendar.event'].
browse(
e.g def create(self, vals): in project_task model

File: project_event/models/project_task.py [hunk 4]
```diff
@@ -189,12 +194,36 @@ def get_equipment_ids_inside(self):
...
+            self.reservation_event_id)
+        reservation_event.write(
+            {
+                'state': 'cancelled'
+            }
+        )
+
+    @api.multi
+    def draft_resources_reservation(self):
```
Comment (jananjoy) [inline project_event/models/project_task.py:212]: @eilst : draft_resources_reservation(self) not clear what the method is doing [Minor comment]

File: project_resource_calendar/models/calendar_event.py [hunk 5]
```diff
@@ -25,3 +25,11 @@ class CalendarEvent(models.Model):
         comodel_name='resource.calendar.instrument',
         ondelete='set null',
     )
+    state = fields.Selection([
+        ('draft', 'Unconfirmed'),
+        ('open', 'Confirmed'),
+        ('cancelled', 'Cancelled')],
```
Comment (jananjoy) [inline project_resource_calendar/models/calendar_event.py:31]: @eilst :" ]," in the new line , Check the example in Instrument.py

File: project_resource_calendar/models/calendar_event.py [hunk 6]
```diff
@@ -25,3 +25,11 @@ class CalendarEvent(models.Model):
...
+        string='Status',
+        readonly=True,
+        track_visibility='onchange',
+        default='draft')
```
Comment (jananjoy) [inline project_resource_calendar/models/calendar_event.py:35]: @eilst : Check the example in Instrument.py for formatting the selection field

File: project_event/models/project_project.py [hunk 7]
```diff
@@ -30,6 +30,30 @@ class Project(models.Model):
     notes = fields.Text(
         string='Notes',
     )
+    state = fields.Selection([
```
Comment (jananjoy) [inline project_event/models/project_project.py]: @eilst :     _description = __doc__  can be included in the class header

File: project_event/models/project_project.py [hunk 8]
```diff
@@ -30,6 +30,30 @@ class Project(models.Model):
...
+        ('draft', 'Draft'),
```
Comment (jananjoy) [inline project_event/models/project_project.py]: @eilst : string='Number', can be included for code field

File: project_event/models/project_task.py [hunk 9]
```diff
@@ -102,6 +102,9 @@ class Task(models.Model):
         default='draft',
         track_visibility='onchange',
     )
+    reservation_event_id = fields.Integer(
```
Comment (jananjoy) [inline project_event/models/project_task.py:104]: @eilst : required=True, why name is required fields. If Yes why name field in project.project is not set required . Please apply the same comments where applicable

File: project_event/models/project_project.py (hunk 8 above)
Comment (jananjoy) [inline project_event/models/project_project.py]: @eilst : why store=true for responsible_id . It is not computed field too

File: project_event/models/project_task.py [hunk 10]
```diff
@@ -102,6 +102,9 @@ class Task(models.Model):
...
+        string='Reservation event',
+    )
 
     @api.onchange('resource_type')
```
Comment (jananjoy) [inline project_event/models/project_task.py:108]: @eilst : Whatis the purpose of setting to "None" on onchange of resource type. Not clear with this method usability

File: project_event/models/project_task.py [hunk 11]
```diff
@@ -171,8 +174,10 @@ def request_reservation(self):
             'resource_type': self.resource_type,
             'room_id': self.room_id.id if self.room_id else None,
             'equipment_ids': [(4, self.equipment_id.id, 0)] if self.equipment_id else None,
+            'state': 'open'
         }
         new_event = calendar_event.create(values)
+        self.reservation_event_id = new_event.id
         if self.room_id:
```
Comment (jananjoy) [inline project_event/models/project_task.py:182]: @eilst:  if self.room_id:
            self.reserve_equipment_inside(new_event.id
If this condition fails, how the application will behave or this condition execute always true

File: project_event/views/project_project_view.xml [hunk 12]
```diff
@@ -26,6 +26,12 @@
         <field name="model">project.project</field>
         <field name="arch" type="xml">
             <form string="Project Event">
+            <header>
+                    <button name="action_option" string="Option" attrs="{'invisible': [('state', '!=', 'draft')]}" type="object"/>
+                    <button name="action_accept" string="Accept" attrs="{'invisible': [('state', 'in', ['accepted','canceled'])]}" type="object"/>
+                    <button name="action_cancel" string="Cancel" type="object" attrs="{'invisible': [('state', 'in', ['canceled'])]}"/>
+                    <field name="state" widget="statusbar" statusbar_visible="draft,option,accepted,canceled"/>
+                </header>
```
Comment (jananjoy) [inline project_event/views/project_project_view.xml:35]: @eilst : Look goods but you are working on the old version of the files. Take  the latest code when you are updating the comments

File: project_event/models/project_project.py (hunk 8 above)
Comment (eilst) [reply project_event/models/project_project.py]: Done

File: project_event/models/project_project.py (hunk 7 above)
Comment (eilst) [reply project_event/models/project_project.py]: Done

File: project_event/models/project_project.py (hunk 1 above)
Comment (eilst) [reply project_event/models/project_project.py]: It's already on chronological order. I guess you meant alphabetical order, but in this case it's better chronological order

File: project_event/models/project_project.py (hunk 7 above)
Comment (rimbendhaou) [inline project_event/models/project_project.py]: @eilst Why you add task states in project.project ?

File: project_event/models/project_project.py (hunk 1 above)
Comment (rimbendhaou) [reply project_event/models/project_project.py]: @eilst Which task?

File: project_event/models/project_task.py (hunk 10 above)
Comment (eilst) [reply project_event/models/project_task.py:108]: To oly have one kind of resource

File: project_event/views/project_project_view.xml (hunk 12 above)
Comment (eilst) [reply project_event/views/project_project_view.xml:35]: well it depends, if my PR is merged before yours.. still the current version

File: project_event/models/project_task.py (hunk 4 above)
Comment (eilst) [reply project_event/models/project_task.py:212]: it is drafting

File: project_event/models/project_task.py [hunk 13]
```diff
@@ -153,12 +156,12 @@ def action_done(self):
...
+        self.draft_resources_reservation()
         self.write({'task_state': 'requested'})
 
     @api.multi
     def action_option(self):
-        self.request_reservation()
+        self.draft_resources_reservation()
         self.write({'task_state': 'option'})
 
     @api.multi
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:167]: @eilst self.ensure_one()

File: project_event/models/project_task.py [hunk 14]
```diff
@@ -171,8 +174,10 @@ def request_reservation(self):
             'resource_type': self.resource_type,
             'room_id': self.room_id.id if self.room_id else None,
             'equipment_ids': [(4, self.equipment_id.id, 0)] if self.equipment_id else None,
+            'state': 'open'
         }
         new_event = calendar_event.create(values)
+        self.reservation_event_id = new_event.id
         if self.room_id:
             self.reserve_equipment_inside(new_event.id)
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:184]: @eilst self.ensure_one()

File: project_event/models/project_task.py (hunk 2 above)
Comment (rimbendhaou) [inline project_event/models/project_task.py]: @eilst if no reservation_event?

File: project_event/models/project_task.py [hunk 15]
```diff
@@ -189,12 +194,36 @@ def get_equipment_ids_inside(self):
         room_id = self.env['resource.calendar.room'].browse(self.room_id).id
         return room_id.instruments_ids.ids
 
+    @api.multi
+    def cancel_resources_reservation(self):
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:200]: @eilst self.ensure_one()

File: project_event/models/project_task.py (hunk 4 above)
Comment (rimbendhaou) [inline project_event/models/project_task.py:212]: @eilst self.ensure_one()

File: project_event/views/project_project_view.xml [hunk 16]
```diff
@@ -26,6 +26,12 @@
         <field name="model">project.project</field>
         <field name="arch" type="xml">
             <form string="Project Event">
+            <header>
+                    <button name="action_option" string="Option" attrs="{'invisible': [('state', '!=', 'draft')]}" type="object"/>
+                    <button name="action_accept" string="Accept" attrs="{'invisible': [('state', 'in', ['accepted','canceled'])]}" type="object"/>
+                    <button name="action_cancel" string="Cancel" type="object" attrs="{'invisible': [('state', 'in', ['canceled'])]}"/>
```
Comment (rimbendhaou) [inline project_event/views/project_project_view.xml]: @eilst add action return to draft

**Review by rimbendhaou — CHANGES_REQUESTED** [review]

File: project_event/models/project_task.py (hunk 13 above)
Comment (eilst) [reply project_event/models/project_task.py:167]: done

File: project_event/models/project_task.py (hunk 14 above)
Comment (eilst) [reply project_event/models/project_task.py:184]: ok, done

File: project_event/models/project_task.py (hunk 15 above)
Comment (eilst) [reply project_event/models/project_task.py:200]: done

File: project_event/models/project_task.py (hunk 2 above)
Comment (eilst) [reply project_event/models/project_task.py]: It is a cancel method, so for sure there most be one event. Should we add the exception anyway?

File: project_event/models/project_task.py (hunk 4 above)
Comment (eilst) [reply project_event/models/project_task.py:212]: done

File: project_event/views/project_project_view.xml [hunk 17]
```diff
@@ -26,6 +26,13 @@
         <field name="model">project.project</field>
         <field name="arch" type="xml">
             <form string="Project Event">
+            <header>
+                    <button name="action_option" string="Option" attrs="{'invisible': [('state', '!=', 'draft')]}" type="object"/>
+                    <button name="action_accept" string="Accept" attrs="{'invisible': [('state', 'in', ['accepted','canceled'])]}" type="object"/>
+                    <button name="action_draft" string="Return to Draft" attrs="{'invisible': [('state', 'not in', ['accepted'])]}" type="object"/>
```
Comment (rimbendhaou) [inline project_event/views/project_project_view.xml]: @eilst invisible if state in draft or canceled

File: project_event/views/project_project_view.xml [hunk 18]
```diff
@@ -26,6 +26,13 @@
...
+                    <button name="action_cancel" string="Cancel" type="object" attrs="{'invisible': [('state', 'in', ['canceled'])]}"/>
```
Comment (rimbendhaou) [inline project_event/views/project_project_view.xml]: @eilst [('state', '!=', 'canceled')]

File: project_event/views/project_task_view.xml [hunk 19]
```diff
@@ -6,6 +6,12 @@
         <field eval="1" name="priority"/>
         <field name="arch" type="xml">
             <form string="Activity">
+             <header>
+                    <button name="action_option" string="Option" attrs="{'invisible': [('task_state', '!=', 'draft')]}" type="object"/>
+                    <button name="action_accept" string="Accept" attrs="{'invisible': [('task_state', 'in', ['accepted','canceled'])]}" type="object"/>
+                    <button name="action_cancel" string="Cancel" type="object" attrs="{'invisible': [('task_state', 'in', ['done','canceled'])]}"/>
```
Comment (rimbendhaou) [inline project_event/views/project_task_view.xml]: @eilst same thing for activity you need to add return to draft

Comment (rimbendhaou) [reply project_event/views/project_task_view.xml]: @eilst the workflow of activity and task are similar

File: project_event/models/project_task.py [hunk 20]
```diff
@@ -153,16 +155,17 @@ def action_done(self):
 
     @api.multi
     def action_request(self):
-        self.request_reservation()
+        self.draft_resources_reservation()
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:158]: @eilst we create a calendar event if the project.task type is activity?

**Review by rimbendhaou — CHANGES_REQUESTED** [review]

File: project_event/models/project_task.py (hunk 2 above)
Comment (rimbendhaou) [reply project_event/models/project_task.py]: @eilst we can cancel a draft task with no reservation

**Review by rimbendhaou — CHANGES_REQUESTED** [review]
add the translation also

**Review by rimbendhaou — CHANGES_REQUESTED** [review]
@eilst add translation please

**Pull request closed without merging**